### PR TITLE
dax onboarding and dialogs

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/392891325557410/1162375367527169
Tech Design URL:
CC:

**Description**:

Adds Dax Onboarding and Dialogs

_Note: launch with "mb" variant set._

**Steps to test this PR**:

***User Flow Test***:
1. Clean install
1. Launch -> Should show dax onboarding followed by add to home row 
1. First visit to home screen -> Should show first home screen dialog
1. Search for something -> Should see search browsing dialog
1. Go to wikipedia -> No trackers message
1. Go to website with trackers - Should "I blocked them" message
1. Fire button or new tab
1. Should see second home screen dialog "You've got this"
1. 
a) Go to Facebook.com -> Should see major tracker message

**OR**

b) Go to Instagram.com -> Should see owned by major tracker message

***Alternative Tracker Messages***:

If you replace the `DefaultDaxOnboardingSettings` with `InMemoryDaxOnboardingSettings` in the `DaxDialogs` class then it won't persist and you can test each message state:

* 1 tracker 
* 2 trackers
* &gt; 2 trackers

***Variants***:

* Ensure app is allocated to `ma` or `mb`.
* In `ma` variant previous onboarding and tooltips should show.
* From an upgrade ensure no allocation to a variant.

**Orientation Testing**:

* [x] Portrait
* [x] Landscape

**Device Testing**:

* [x] iPhone SE (1st Gen)
* [x] iPhone 8
* [x] iPhone X
* [x] iPad

**OS Testing**:

* [x] iOS 10
* [x] iOS 11
* [x] iOS 12
* [x] iOS 13

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Core/UIViewExtension.swift [hunk 1]
```diff
@@ -66,7 +66,6 @@ extension UIView {
         layer.shadowOffset = CGSize(width: 0, height: 1.5)
         layer.shadowOpacity = 0.2
         layer.shadowRadius = 1.5
-        layer.shadowPath = UIBezierPath(rect: bounds).cgPath
```
Comment (brindy) [inline Core/UIViewExtension.swift:69]: Doesn't round corners properly.  Removing this creates the right shadow anyway - didn't appear to have any negative effects where it was used elsewhere.

File: IntegrationTests/AtbIntegrationTests.swift [hunk 2]
```diff
@@ -60,7 +60,8 @@ class AtbIntegrationTests: XCTestCase {
         app.launchEnvironment = [
             "BASE_URL": "http://localhost:8080",
             "BASE_PIXEL_URL": "http://localhost:8080",
-            "VARIANT": Variant.defaultVariants[0].name // just has to match an existing variant to prevent one being allocated and written to storage
+            // usually just has to match an existing variant to prevent one being allocated and written to storage but we're changing oboarding
+            "VARIANT": "ma"
```
Comment (brindy) [inline IntegrationTests/AtbIntegrationTests.swift:64]: These tests require legacy onboarding.

**Comment by brindy** [comment]
@bwaresiak  I've had some feedback around the logic of when the dialogs display that I need to fix so feel free to hold fire on this until I push a fix.  It won't be a huge change though if you wanted to start reviewing.  Up to you.

**Comment by brindy** [comment]
@bwaresiak  this is ready for review when you get chance. Thanks!

**Comment by brindy** [comment]
@bwaresiak few updates since last time but the product review is basically approved now so that should be it.

**Comment by brindy** [comment]
@bwaresiak few updates based on discussion yesterday.

File: DuckDuckGo/TabViewController.swift [hunk 3]
```diff
@@ -224,7 +224,9 @@ class TabViewController: UIViewController {
     }
     
     func installBrowsingTips() {
-        tips = BrowsingTips(delegate: self)
+        if DefaultVariantManager().isSupported(feature: .legacyOnboarding) {
```
Comment (bwaresiak) [inline DuckDuckGo/TabViewController.swift]: Maybe it should be the other way around:
`if DefaultVariantManager().isSupported(feature: . daxOnboarding) {` 
so after migration we will still show any pending onboarding tips?

File: DuckDuckGo/FullscreenDaxDialogViewController.swift [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    
+    weak var daxDialogViewController: DaxDialogViewController?
+    weak var delegate: FullscreenDaxDialogDelegate?
+
+    var spec: DaxDialogs.BrowsingSpec?
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        daxDialogViewController?.applyTheme(ThemeManager.shared.currentTheme)
```
Comment (bwaresiak) [inline DuckDuckGo/FullscreenDaxDialogViewController.swift]: This should be encapsulated in Dax's VC `viewDidLoad()` - calling it from the outside gives no guarantee that view has been already loaded.

**Comment by brindy** [comment]
@bwaresiak have updated based on your comments - thanks!

File: DuckDuckGo/TabViewController.swift [hunk 5]
```diff
@@ -773,18 +785,33 @@ extension TabViewController: WKNavigationDelegate {
...
+            let siteRating = self.siteRating,
+            let spec = DaxDialogs().nextBrowsingMessage(siteRating: siteRating) else {
+                scheduleTrackerNetworksAnimation(collapsing: true)
+                return
+        }
+        
+        scheduleTrackerNetworksAnimation(collapsing: !spec.highlightAddressBar)
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
+            self?.chromeDelegate?.setBarsHidden(false, animated: true)
+            self?.performSegue(withIdentifier: "DaxDialog", sender: spec)
```
Comment (bwaresiak) [inline DuckDuckGo/TabViewController.swift:808]: I'd consider adding `self?.chromeDelegate?.omniBar.resignFirstResponder()` to ensure that tracker logos are visible when presenting first dax onboarding message.

**Review by bwaresiak — APPROVED** [review]
Great job @brindy , LGTM :)

**Pull request merged**